File: extensions/syncforreddit/src/main/java/app/morphe/extension/syncforreddit/ui/lightbox/TransparentAlphaPatch.java
package app.morphe.extension.syncforreddit.ui.lightbox;

import android.graphics.Bitmap;
import android.graphics.Color;

import app.morphe.extension.shared.Logger;
import app.morphe.extension.syncforreddit.settings.SyncUpSettings;

/**
 * What a picture with transparency in it is drawn as in the lightbox.
 *
 * <p>The zooming view Sync opens a picture in decodes through Skia, and where nothing says
 * otherwise that decoder is built with {@code RGB_565} — a format with no alpha channel at all.
 * Everything transparent therefore arrives opaque, showing whatever was underneath it in the
 * file, which for most pictures saved with a matte is white.
 *
 * <p>Naming a preferred config is the library's own way of settling this, and the decoders read
 * it before falling back, so answering here is enough for every one of them.
 *
 * @noinspection unused
 */
public final class TransparentAlphaPatch {
    private static final String KEEP_ALPHA = "sync_up_lightbox_alpha";

    /** Reported once rather than for every tile a large picture is decoded in. */
    private static boolean said;

    private TransparentAlphaPatch() {}

    public static boolean isPatchIncluded() {
        // Overridden by patch.
        return false;
    }

    /**
     * @param original What the view was going to decode as, which is null unless something has
     *                 already asked for a particular format.
     * @return The format to decode a picture in the lightbox as.
     */
    public static Bitmap.Config preferredConfig(Bitmap.Config original) {
        if (!isPatchIncluded()) {
            return original;
        }

        boolean keep = SyncUpSettings.flag(KEEP_ALPHA, true);
        if (!said) {
            said = true;
            SyncUpSettings.report("lightbox", KEEP_ALPHA);
            Logger.printDebug(() -> "lightbox: asked what to decode as, was " + original
                    + ", keep alpha is " + keep);
        }

        if (!keep) {
            return original;
        }
        // Only where nothing else has chosen: a caller that named a format meant it.
        if (original != null && original != Bitmap.Config.RGB_565) {
            Logger.printDebug(() -> "lightbox: leaving the chosen " + original + " alone");
            return original;
        }
        Logger.printDebug(() -> "lightbox: decoding as ARGB_8888 so transparency survives");
        return Bitmap.Config.ARGB_8888;
    }

    /**
     * What the zooming view paints behind each tile of the picture.
     *
     * <p>Decoding with an alpha channel is only half of it: the view fills every tile with a
     * colour first, and the viewer asks for opaque white. Transparency therefore arrived intact
     * and was then painted over, which is why the preview above a post was right while the same
     * picture opened was not.
     *
     * @param original The colour the viewer asked for.
     * @return The colour to paint behind the picture.
     */
    public static int tileBackground(int original) {
        if (!isPatchIncluded() || !SyncUpSettings.flag(KEEP_ALPHA, true)) {
            return original;
        }
        if (original == Color.TRANSPARENT) {
            return original;
        }
        Logger.printDebug(() -> "lightbox: painting nothing behind the tiles rather than "
                + String.format("#%08X", original));
        return Color.TRANSPARENT;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/reddit/customclients/sync/syncforreddit/ui/lightbox/TransparentAlphaPatch.kt
package app.morphe.patches.reddit.customclients.sync.syncforreddit.ui.lightbox

import app.morphe.patcher.Fingerprint
import app.morphe.patcher.extensions.InstructionExtensions.addInstructions
import app.morphe.patcher.extensions.InstructionExtensions.getInstruction
import app.morphe.patcher.extensions.InstructionExtensions.instructions
import app.morphe.patcher.patch.PatchException
import app.morphe.patcher.patch.bytecodePatch
import app.morphe.patches.reddit.customclients.sync.SyncForRedditCompatible
import app.morphe.patches.reddit.customclients.sync.syncforreddit.extension.sharedExtensionPatch
import app.morphe.util.returnEarly
import com.android.tools.smali.dexlib2.Opcode
import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction

private const val EXTENSION_CLASS_DESCRIPTOR =
    "Lapp/morphe/extension/syncforreddit/ui/lightbox/TransparentAlphaPatch;"

private const val PREFERRED_CONFIG_METHOD =
    "preferredConfig(Landroid/graphics/Bitmap\$Config;)Landroid/graphics/Bitmap\$Config;"

private const val TILE_BACKGROUND_METHOD = "tileBackground(I)I"

/** The zooming view a picture is opened in, which keeps the format to decode in statically. */
private const val ZOOMING_VIEW = "Lcom/davemorrissey/labs/subscaleview/SubsamplingScaleImageView;"

@Suppress("unused")
val transparentAlphaPatch = bytecodePatch(
    name = "Keep transparency in a picture",
    description = "Draws the transparent parts of a picture opened in the viewer as the " +
            "background behind it rather than as white.",
    default = true
) {
    dependsOn(sharedExtensionPatch, transparentAlphaSettingsPatch)

    compatibleWith(*SyncForRedditCompatible)

    execute {
        Fingerprint(
            definingClass = EXTENSION_CLASS_DESCRIPTOR,
            name = "isPatchIncluded",
        ).method.returnEarly(true)

        // The decoders ask the view what to decode in and fall back to RGB_565, which has no
        // alpha channel, when it answers nothing. Answering here covers every one of them.
        Fingerprint(
            definingClass = ZOOMING_VIEW,
            name = "getPreferredBitmapConfig",
            parameters = listOf(),
            returnType = "Landroid/graphics/Bitmap\$Config;",
        ).method.apply {
            val returnIndex = instructions.indexOfFirst { it.opcode == Opcode.RETURN_OBJECT }
            if (returnIndex < 0) {
                throw PatchException("The zooming view never answers what to decode in")
            }
            val register = getInstruction<OneRegisterInstruction>(returnIndex).registerA

            addInstructions(
                returnIndex,
                """
                invoke-static       { v$register }, $EXTENSION_CLASS_DESCRIPTOR->$PREFERRED_CONFIG_METHOD
                move-result-object  v$register
                """
            )
        }

        // The view fills every tile with a colour before drawing the picture over it, and the
        // viewer asks for opaque white, so transparency survived the decode and was painted
        // over anyway.
        Fingerprint(
            definingClass = ZOOMING_VIEW,
            name = "setTileBackgroundColor",
            parameters = listOf("I"),
            returnType = "V",
        ).method.addInstructions(
            0,
            """
            invoke-static       { p1 }, $EXTENSION_CLASS_DESCRIPTOR->$TILE_BACKGROUND_METHOD
            move-result         p1
            """
        )
    }
}
